backend/src/github-util.ts

```diff
@@ -0,0 +1,242 @@
+  if (match == null) {
+    throw new Error(`Malformed URL ${url}.`); // more helpful error message?
+  }
+
+  return { owner: match[1], repo: match[2], pull_number: parseInt(match[3], 10) };
+};
+
+/** Retrieves comments made on `pull_request`. */
+const getReviewComments = async (pull_request: PullRequest): Promise<ReviewComment[]> => {
+  const octokit = new Octokit();
+
+  return octokit.rest.pulls.listReviewComments(pull_request).then((res) => {
+    // handle non 200 response?
+
+    const prs = res.data;
+    return prs.map(
+      (pr): ReviewComment => ({
+        comment_url: pr.html_url,
+        created_by: pr.user.login,
+        created_at: Date.parse(pr.created_at),
+        content: pr.body
+      })
+    );
+  });
+};
+
+/** Returns `comments` created by `username` between `start_time` and `end_time`.
+ *  Raises an error if no comment satisfies these conditions.
+ */
+const filterComments = (
+  comments: ReviewComment[],
+  username: string,
+  start_time: number,
+  end_time: number
+): ReviewComment[] => {
+  if (!comments || !comments.length) {
+    throw new Error(`No review comments in this PR.`);
+  }
+
+  let eligible_comments: ReviewComment[] = [];
+
+  // comments made by user
+  eligible_comments = comments.filter((comment) => comment.created_by === username);
+  if (!eligible_comments || !eligible_comments.length) {
+    throw new Error(`No review comments made by user ${username} in this PR.`);
+  }
+
+  // comments made in date range
+  eligible_comments = eligible_comments.filter(
+    (comment) => start_time <= comment.created_at && comment.created_at <= end_time
+  );
+  if (!eligible_comments || !eligible_comments.length) {
+    const start_date = new Date(start_time).toDateString();
+    const end_date = new Date(end_time).toDateString();
+    throw new Error(
+      `No review comments made by user ${username} between ${start_date} and ${end_date}.`
+    );
+  }
+
+  return eligible_comments;
+};
+
+/** Retrieves information about `pull_request` and its review comments. */
+const getReviewedPR = async (pull_request: PullRequest): Promise<ReviewedPR> => {
+  const octokit = new Octokit();
+
+  // get information about a PR and its review comments
+  // cannot get both with a single api call
+  return Promise.all([octokit.rest.pulls.get(pull_request), getReviewComments(pull_request)]).then(
+    ([pr, comments]) => ({
+      url: pr.data.html_url,
+      created_by: pr.data.user?.login || '',
+      comments
+    })
+  );
+};
+
+/** Retrieves information about opened PR `pull_request`. */
+const getOpenedPR = async (pull_request: PullRequest): Promise<OpenedPR> => {
+  const octokit = new Octokit();
+
+  return octokit.rest.pulls.get(pull_request).then((res) => {
+    // handle non 200 response?
+
+    const pr = res.data;
+    return {
+      url: pr.html_url,
+      created_by: pr.user?.login || '',
+      created_at: Date.parse(pr.created_at),
+      diff_size: pr.additions + pr.deletions
```

Comment:
we might want to consider getting this info from the big diff bot -- i believe it does the calculation in a slightly smarter way that isn't just additions + deletions

Reply:
> we might want to consider getting `diff_size` from the big diff bot -- i believe it does the calculation in a slightly smarter way that isn't just additions + deletions

While implementing this, I realized that Github treats thread comments (like the one im writing now) differently than comments made for a review (like the one im quoting). While the PR requirement is for "reviews," should we take both review comments and non-review comments into consideration? It seems like a technicality if someone's review PR is determined to be invalid just because they made only "regular" comments instead of review comments.